src/simulation/LoadGenerator.cpp

```diff
@@ -221,591 +143,459 @@ maybeAdjustRate(double target, double actual, uint32_t& rate, bool increaseOk)
+LoadGenerator::generateLoad(bool isCreate, uint32_t nAccounts, uint32_t nTxs,
+                            uint32_t txRate, uint32_t batchSize, bool autoRate)
 {
-    soci::transaction sqltx(app.getDatabase().getSession());
-    app.getDatabase().setCurrentTransactionReadOnly();
+    soci::transaction sqltx(mApp.getDatabase().getSession());
+    mApp.getDatabase().setCurrentTransactionReadOnly();
+    getRootAccount();
 
-    updateMinBalance(app);
+    // Finish if no more txs need to be created.
+    if ((isCreate && nAccounts == 0) || (!isCreate && nTxs == 0))
+    {
+        // Done submitting the load, now ensure it propagates to the DB.
+        waitTillComplete();
+        return;
+    }
 
+    updateMinBalance();
     if (txRate == 0)
     {
         txRate = 1;
     }
-
-    // txRate is "per second"; we're running one "step" worth which is a
-    // fraction of txRate determined by STEP_MSECS. For example if txRate
-    // is 200 and STEP_MSECS is 100, then we want to do 20 tx per step.
-    uint32_t txPerStep = (txRate * STEP_MSECS / 1000);
-
-    // There is a wrinkle here though which is that the tx-apply phase might
-    // well block timers for up to half the close-time; plus we'll be probably
-    // not be scheduled quite as often as we want due to the time it takes to
-    // run and the time the network is exchanging packets. So instead of a naive
-    // calculation based _just_ on target rate and STEP_MSECS, we also adjust
-    // based on how often we seem to be waking up and taking loadgen steps in
-    // reality.
-    auto& stepMeter =
-        app.getMetrics().NewMeter({"loadgen", "step", "count"}, "step");
-    stepMeter.Mark();
-    auto stepsPerSecond = stepMeter.one_minute_rate();
-    if (stepMeter.count() > 10 && stepsPerSecond != 0)
+    if (batchSize == 0)
     {
-        txPerStep = static_cast<uint32_t>(txRate / stepsPerSecond);
+        batchSize = 1;
     }
 
-    // If we have a very low tx rate (eg. 2/sec) then the previous division will
-    // be zero and we'll never issue anything; what we need to do instead is
-    // dispatch 1 tx every "few steps" (eg. every 5 steps). We do this by random
-    // choice, weighted to the desired frequency.
-    if (txPerStep == 0)
-    {
-        txPerStep = rand_uniform(0U, 1000U) < (txRate * STEP_MSECS) ? 1 : 0;
-    }
+    uint32_t txPerStep = getTxPerStep(txRate);
+    auto& submitTimer =
+        mApp.getMetrics().NewTimer({"loadgen", "step", "submit"});
+    auto submitScope = submitTimer.TimeScope();
 
-    if (txPerStep > nTxs)
-    {
-        // We're done.
-        CLOG(INFO, "LoadGen") << "Load generation complete.";
-        app.getMetrics().NewMeter({"loadgen", "run", "complete"}, "run").Mark();
-        clear();
-    }
-    else
+    uint32_t ledgerNum = mApp.getLedgerManager().getLedgerNum();
+    uint64_t numToProcess = 1;
+    uint64_t totalAccounts = mAccounts.size();
+
+    for (uint32_t i = 0; i < txPerStep; ++i)
```

Comment:
It looks for me as if there are 2 different methods hiding inside this loop, one for isCreate == true and one for == false. Or maybe it is even small interface with 2 implementation. Following all those if (isCreate) branches one after another is quite hard.